Fix 'oversized' caveat

milljm opened the pull request:
getCommand was clobbering any previously set
caveats.

Scheduler was only taking into account CPUs and
not a combination of CPUs and Threads when
performing slot availability.

Add unittests to verify resource allocation caveats
are generated when they should be.

Closes #10272

Commit:
Fix 'oversized' caveat

getCommand was clobbering any previously set
caveats.

Scheduler was only taking into account CPUs and
not a combination of CPUs and Threads when
performing slot availability.

Add unittests to verify resource allocation caveats
are generated when they should be.

Closes #10272

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/135505/) on 18ae4de wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10284/site)

This comment will be updated on new commits.

permcody commented on python/TestHarness/testers/Tester.py:
```diff
@@ -312,6 +313,10 @@ def getProcs(self, options):
 
+    def getSlots(self, options):
+        """ return number of slots to use for this tester """
+        return int(self.getThreads(options)) * int(self.getProcs(options))
```
Why you are you casting here? These methods should both be returning ints already.

milljm replied:
Just being carful, because getThreads/Procs in the base class is casted. But perhaps the derived class is dumb..?  I'll change it to satisfy this comment.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -362,11 +362,8 @@ def canPrint(self, tester):
-
-        if tester.specs.isValid('caveats'):
-            caveats = tester.specs['caveats']
+        caveats = list(tester.getCaveats())
```
Hmm, I suppose we can allow this but you really should follow up and try to clean this up everywhere. I'm not a huge fan of testing against python types to do different things. I'll comment on that below.

permcody commented on python/TestHarness/testers/Tester.py:
```diff
@@ -397,6 +402,18 @@ def getRedirectedOutputFiles(self, options):
 
+    def addCaveats(self, caveats):
+        """ Add caveat(s) which will be displayed with the final test status """
+        if type(caveats) == type([]):
```
Not a huge fan of this line. How much word would it be to just make sure that caveats is always a list?

permcody reviewed (commented):
This is mergeable, but it creates a few new things that will need to be cleaned up. I'd prefer you attempt to add one more commit to see if you can clean up caveats everywhere. Nice testing!

Commit:
Address comments

Don't check for type, but rather use non-named expansion (*kwargs)
when adding caveats.

Remove type casting when determining slots.

Modify existing methods that can make use of new tester caveat
methods.

Adding a unit test to handle all the 'check' caveats that I found
that were not being printed when they should have been.

Closes #10272

milljm commented:
I forgot to address TestHarness formatCaveats... do not merge...

milljm commented on test/tests/test_harness/extra_info:
```diff
@@ -0,0 +1,27 @@
+    asio = 'TODO'
+    boost = 'TODO'
+  [../]
+[]
```
Even though they say 'TODO' they do work... I forgot to change the name. Well, actually I didn't know what to put in all those fields... true? A version number? Anyways, it doesn't really matter. The caveat _is_ being displayed.

permcody replied:
Hmm...  Looks like we need to start cleaning these up! We don't need cxx11 anymore for instance. Not sure about asio.

permcody replied:
I really think we should fill these in with values... I know you don't want to stretch this out, but if I merge this, it'll probably be in there for awhile. What I'd do is search for "tests" files and grep for each one of these. It seems like it won't take long to fill them in. After that, we'll clean them up (another ticket).

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -337,7 +337,7 @@ def appendRecoverableTests(self, testers):
                 part2_params['delete_output_before_running'] = False
                 part2_params['cli_args'].append('--recover --recoversuffix ' + self.options.recoversuffix)
-                part2_params.addParam('caveats', ['recover'], "")
+                part2.addCaveats('recover')
```
TODO: Make a unittest for this caveat.

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -360,32 +360,29 @@ def canPrint(self, tester):
-            if len(caveats):
-                result = '[' + ', '.join(caveats).upper() + '] ' + tester.getSuccessMessage()
+                for check in self.options._checks.keys():
+                    if tester.specs.isValid(check) and not 'ALL' in tester.specs[check]:
```
we can do the same in the Tester.py checkRunnable method?

milljm commented on python/TestHarness/TestHarness.py:
```diff
@@ -360,32 +360,29 @@ def canPrint(self, tester):
+                        tester.addCaveats(check)
+            if tester.getCaveats():
+                result = '[' + ', '.join(tester.getCaveats()).upper() + '] ' + tester.getSuccessMessage()
             else:
```
We can perhaps do the same thing in Tester.py checkRunnable...
